File: chunlin/139.java
class Solution {
    public boolean wordBreak(String s, List<String> wordDict) {
        boolean[] isSegmentOk = new boolean[s.length() + 1];
        isSegmentOk[0] = true; // none of char
            for (int i = 1; i < s.length() + 1; i++) {
                int j = i - 1;
                while (j >= 0) {
                    if (isSegmentOk[j] == true && wordDict.contains(s.substring(j, i))) {
                        isSegmentOk[i] = true;
                        break;
                    } else {
                        j--;
                    }
                }
            }
        return isSegmentOk[s.length()];
    }
}

File: zemiao/139.py
class Solution(object):
    def wordBreak(self, s, wordDict):
        """
        :type s: str
        :type wordDict: List[str]
        :rtype: bool
        """
        tried = {}
        
        def dfs(s):
            if s == "":
                return True
            if s in tried:
                return False
            
            n = len(s)
            for w in wordDict:
                if len(w) <= n and w == s[:len(w)] and dfs(s[len(w):]):
                    return True
                
            tried[s] = True
            return False
        
        return dfs(s)